system/mp/mpcomm.cpp

```diff
@@ -2945,46 +2949,99 @@ class CCommunicator: public ICommunicator, public CInterface
-                    doverify = (myrank>rank);
-                else
-                    doverify = (myrank<rank);
-                if (doverify) {
-                    Owned<CMPChannel> channel = getChannel(rank);
+            CTimeMon tm(timeout);
+            Owned<CMPChannel> channel = getChannel(rank);
+            return channel->verifyConnection(tm, true);    
+        };
+
+        auto verifyConnWithoutConnect = [&](unsigned rank, unsigned timeout)
+        {
+            CTimeMon tm(timeout);
+            while (true)
+            {
+                Owned<CMPChannel> channel = getChannel(rank);
+                if (channel->verifyConnection(tm, false))
+                    return true;
+                if (abort || tm.timedout())
+                    return false;
+                Sleep(100);
+            }
+        };
+
+        auto threadedVerifyConnectFunc = [&](rank_t rank, std::function<bool (unsigned rank, unsigned timeout)> connectFunc)
+        {
+            // NB: running because took (via wait()) a semaphore slot, restore it at end of scope
+            struct RestoreSlot
+            {
+                Semaphore &sem;
+                RestoreSlot(Semaphore &_sem) : sem(_sem) { }
+                ~RestoreSlot() { sem.signal(); }
+            } restoreSlot(sem);
+
+            unsigned timeoutMs;
+            if (totalTM.timedout(&timeoutMs) || abort)
+                return false;
+            if (perConnectionTimeout && (perConnectionTimeout < timeoutMs))
+                timeoutMs = perConnectionTimeout;
+
+            if (!connectFunc(rank, timeoutMs))
+            {
+                abort = true; // ensure verifyFunc knows before release slot, to prevent other thread being launched
+                return false;
+            }
+            return true;
+        };
+
+        auto verifyFunc = [&](std::function<bool (unsigned rank)> conditionFunc, std::function<bool (unsigned rank, unsigned timeout)> connectFunc)
+        {
+            std::vector<std::future<bool>> threads;
+            for (rank_t rank=0; rank<group->ordinality(); rank++)
+            {
+                if (conditionFunc(rank))
+                {
+                    // check timeout before and after sem.wait
+                    // NB: sem.wait if successful, takes a slot which is restored by the thread when it is done
                     unsigned remaining;
-                    if (tm.timedout(&remaining)) {
-                        return false;
+                    if (totalTM.timedout(&remaining) || !sem.wait(remaining) || totalTM.timedout(&remaining))
+                    {
+                        abort = true;
+                        break;
                     }
-                    if (!channel->verifyConnection(tm,true))
-                        return false;
+                    else if (abort)
+                        break;
+                    threads.push_back(std::async(threadedVerifyConnectFunc, rank, connectFunc));
```

Comment:
Should this be: ?
threads.push_back(std::async(std::launch::async,threadedVerifyConnectFunc, rank, connectFunc));
Without the std::launch::async I don't think new threads are always created ??

Reply:
not sure, the default behaviour is std::launch::async | std::launch::deferred, which does mean it's implementation dependent, so might limit the number of threads, e.g. if resources are tight.
That may not be a bad thing though.
But I'm not seeing that, I see up to 32 concurrent threads (based on my hardware_concurrency() limit).

Are you not seeing threads created?

I doubt matters much either way here, but given there's a semaphore limiting max here too, it could be changed to std::launch::async.

But I'm curious if you're not seeing thread and why.

Reply:
Correct, no new threads, even with delays added, unless I add std::launch::async.

Reply:
Would it be a compiler version issue ?

Reply:
Or that there are already several threads running and my limit is lower (8) ?  I will try with a forced larger value to see.

Reply:
I tried a larger value (32) and it did not make any difference.

Reply:
I'm not sure why we're seeing very different behviour, I see the max number of threads running in parallel, you see 1.
But as the semantics mean it is guaranteed, and this impl. is self limiting anyway, I will change to std::launch::async